Update wp_parse_url() for performance and legacy code removal

WpSpeedDoctor opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->
Update wp_parse_url() for performance and legacy code removal
With the current supported PHP, it's no longer necessary to use this function's code to mitigate PHP incompatibility. This change aims to remove the legacy fix and improve performance. 

Functions "_get_component_from_parsed_url_array()" and "_wp_translate_php_url_constant_to_key()" are not required for the new version of the function

Trac ticket: https://core.trac.wordpress.org/ticket/62124

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

WpSpeedDoctor committed:
Update wp_parse_url() for performance and legacy code removal

With the current supported PHP, it's no longer necessary to use this function's code to mitigate PHP incompatibility. This change aims to remove the legacy fix and improve performance. 

Functions `_get_component_from_parsed_url_array()` and `_wp_translate_php_url_constant_to_key()` are not required for the new version of the function

WpSpeedDoctor committed:
Merge branch 'WordPress:trunk' into patch-1

WpSpeedDoctor committed:
Update http.php

Update wp_parse_url() for performance and legacy code removal
With the current supported PHP version, this function's code is no longer necessary to mitigate PHP native function incompatibility. This change aims to remove the legacy fix and improve performance. 

Functions `_get_component_from_parsed_url_array()` and `_wp_translate_php_url_constant_to_key()` are not required for the new version of the function, and should be remeoved.

WpSpeedDoctor committed:
Update http.php fixed white spaces

apermo commented on src/wp-includes/http.php:
```diff
@@ -710,31 +710,60 @@ function ms_allowed_http_request_hosts( $is_external, $host ) {
 
-	$parts = parse_url( $url );
+		case -1:
+			return parse_url( (string) $url );
```
You use the typecast `(string)` here and again at the bottom.

I'd either normalize `$url` before the `switch` or simply set `$component = null;` and use the single return at the end to reduce redundancy here.

apermo reviewed (commented):
Looks solid to me, just one comment for concideration.

WpSpeedDoctor committed:
Removed early return

For case -1: was removed return as `-1` of $component is the default value anyway, `null` value is deprecated so it's better to continue the case and return on the end.
I kept typecast `(string) $url` as it's commonly used in WP core code and it's not listed as a bad practice in PHP coding standards. 
Now this should be a performance update, I don't want to declare a new variable like `$string_url=strval($url)` or assign `$url = strval($url);`, from the performance point of view, it's an unnecessary step, this hardly will add to readability.

felixarntz commented on src/wp-includes/http.php:
```diff
@@ -710,31 +710,58 @@ function ms_allowed_http_request_hosts( $is_external, $host ) {
-	}
+		case 'path':
+			$component = PHP_URL_PATH;
+			break;
+
+		case 'query':
+			$component = PHP_URL_QUERY;
+			break;
+
+		case 'scheme':
+			$component = PHP_URL_SCHEME;
+			break;
+
+		case 'host':
+			$component = PHP_URL_HOST;
+			break;
+
+		case 'port':
+			$component = PHP_URL_PORT;
+			break;
+
+		case 'user':
+			$component = PHP_URL_USER;
+			break;
+
+		case 'pass':
+			$component = PHP_URL_PASS;
+			break;
+
+		case 'fragment':
+			$component = PHP_URL_FRAGMENT;
+			break;
 
-	// Remove the placeholder values.
-	foreach ( $to_unset as $key ) {
-		unset( $parts[ $key ] );
+		default:
+			return null;
+			break;
 	}
```
I think more efficient (and more concise to read/edit) than this large `switch` would be to have a static map of `key => constant`, and then we could do a simple `$component = isset( $map[ $component ] ) ? $map[ $component ] : $component`.

The map could essentially be what's currently in `_wp_translate_php_url_constant_to_key()`, just flipped.

WpSpeedDoctor replied:
@felixarntz I disagree with the proposed change to array map and loop to unset. The switch statement is fast, that's why I used it. This is a performance improvement update. Reading this switch is not that hard. Creating an array first and then looping through it may diminish the performance benefits of this update. I understand it's only microseconds, but those count too. I tested it in a real-world scenario, and it does show a speed improvement after the function has been called 30+ times per tested request.

felixarntz commented on src/wp-includes/http.php:
```diff
@@ -710,31 +710,58 @@ function ms_allowed_http_request_hosts( $is_external, $host ) {
+			break;
 	}
 
-	return _get_component_from_parsed_url_array( $parts, $component );
```
This method and `_wp_translate_php_url_constant_to_key()` should probably be deprecated as they're no longer used nor recommended for anything based on this change.

pbearne replied:
@WpSpeedDoctor will you do the deprecation?

WpSpeedDoctor replied:
@pbearne  I would like to clarify that when I created the Trac ticket, I wasn't sure if I could simply delete it or if there was a specific procedure to follow to do it the right way. For me, those can be deprecated.

felixarntz replied:
@WpSpeedDoctor They could be moved to the `deprecated.php` file and receive `@deprecated` annotations as well as `_deprecated_function()` calls inside.

felixarntz reviewed (changes requested):
@WpSpeedDoctor @pbearne Overall this looks reasonable to me, though we should also deprecate the helper functions if we make this change.

For more holistic feedback on the overall change, I'd like to get reviews from either @johnbillion or @peterwilsoncc who worked on `wp_parse_url()` in the past.

WpSpeedDoctor committed:
Removed deprecated functions

https://core.trac.wordpress.org/ticket/62124

WpSpeedDoctor committed:
Moved deprecated functions

Deprecated function from http.php
https://core.trac.wordpress.org/ticket/62124

deprecated functions:

_get_component_from_parsed_url_array( )
_wp_translate_php_url_constant_to_key()

WpSpeedDoctor committed:
Added  @deprecated 6.8.0

WpSpeedDoctor committed:
Deprecated functions tests

https://core.trac.wordpress.org/ticket/62124

WpSpeedDoctor committed:
Whitespace fix

https://core.trac.wordpress.org/ticket/62124

WpSpeedDoctor committed:
Fix white space #2

https://core.trac.wordpress.org/ticket/62124

WpSpeedDoctor committed:
Fix whitespaces #3

https://core.trac.wordpress.org/ticket/62124

WpSpeedDoctor committed:
fix whitespaces, re-run tests

https://core.trac.wordpress.org/ticket/62124

WpSpeedDoctor committed:
Update deprecated.php adjusted WP version of deprecation

The unit test should pass now

peterwilsoncc commented on src/wp-includes/deprecated.php:
```diff
@@ -6424,3 +6424,47 @@ function wp_create_block_style_variation_instance_name( $block, $variation ) {
+ *               doesn't exist in the given URL; a string or - in the case of
+ *               PHP_URL_PORT - integer when it does. See parse_url()'s return values.
+ */
+function _get_component_from_parsed_url_array( $url_parts, $component ) {
```
This and `_wp_translate_php_url_constant_to_key()` should retain their existing code, simply add the deprecation notice at the start of the function. There are a few [plugins that make use of these functions](https://wpdirectory.net/search/01JW4W00DVWARZPVF06H1BZZZ3) even though they are marked as internal. One of which has 4,000,000+ installs.

Note, there are a few false positives in the link above.

WpSpeedDoctor replied:
I've added original code into deprecated functions. But now unit test will fail as it asserts null. I don't know how to roll it back.

peterwilsoncc replied:
If you revert the test changes you made in `tests/phpunit/tests/http/http.php` back to their original form in trunk, they will begin passing again.

The only change you'll need to make to the test suite is to include the `@expectedDeprecated` annotations in the docblocks.

peterwilsoncc commented on tests/phpunit/tests/http/http.php:
```diff
@@ -331,73 +331,46 @@ public function test_normalize_cookies_scalar_values() {
 
 	/**
 	 * @ticket 36356
+	 * @ticket 62124
```
Per my note above, re retaining the code, these tests should remain as is with the addition of the `expectedDeprecated` annotation.

peterwilsoncc commented on src/wp-includes/http.php:
```diff
@@ -710,93 +710,56 @@ function ms_allowed_http_request_hosts( $is_external, $host ) {
  *               PHP_URL_PORT - integer when it does. See parse_url()'s return values.
  */
 function wp_parse_url( $url, $component = -1 ) {
```
Rather than a big long `switch`, it might be easier to do a mapping of values that's the reverse of the array below, ie:


```php
$translation = array(
	'scheme'   => PHP_URL_SCHEME,
	'host'     => PHP_URL_HOST,
	'port'     => PHP_URL_PORT,
	'user'     => PHP_URL_USER,
	'pass'     => PHP_URL_PASS,
	'path'     => PHP_URL_PATH,
	'query'    => PHP_URL_QUERY,
	'fragment' => PHP_URL_FRAGMENT,
);
```

I saw the earlier comment that you were concerned aboput performance, but I don't think there will be much - if any - difference between a switch and the following:

```php
if ( isset( $translation[ $component ] ) ) {
    $component = $translation[ $component ];
}
```

WpSpeedDoctor replied:
I measured it, and the difference is too small. All right, let's do it this way. Thank you!

peterwilsoncc committed:
Merge branch 'trunk' into patch-1

peterwilsoncc reviewed (commented):
I've added some notes inline.

I've taken the liberty of merging trunk in to the PR to resolve the merge conflict in `deprecated.php`

WpSpeedDoctor committed:
switch replaced by array

WpSpeedDoctor committed:
Added the original code into deprecated functions

https://github.com/WordPress/wordpress-develop/pull/7743
